File: src/main/java/io/javalin/core/ExceptionMapper.java
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 *
 */

package io.javalin.core;

import java.util.HashMap;
import java.util.Map;

import javax.servlet.http.HttpServletResponse;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.javalin.ExceptionHandler;
import io.javalin.Request;
import io.javalin.HaltException;
import io.javalin.Response;

public class ExceptionMapper {

    private static Logger log = LoggerFactory.getLogger(ExceptionMapper.class);

    private Map<Class<? extends Exception>, ExceptionHandler> exceptionMap;

    public ExceptionMapper() {
        this.exceptionMap = new HashMap<>();
    }

    public void put(Class<? extends Exception> exceptionClass, ExceptionHandler handler) {
        this.exceptionMap.put(exceptionClass, handler);
    }

    public void clear() {
        this.exceptionMap.clear();
    }


    void handle(Exception e, Request request, Response response) {
        if (e instanceof HaltException) {
            response.status(((HaltException) e).statusCode);
            response.body(((HaltException) e).body);
            return;
        }
        ExceptionHandler exceptionHandler = this.getHandler(e.getClass());
        if (exceptionHandler != null) {
            exceptionHandler.handle(e, request, response);
        } else {
            log.error("Uncaught exception", e);
            response.body("Internal server error");
            response.status(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
        }
    }


    public ExceptionHandler getHandler(Class<? extends Exception> exceptionClass) {
        if (this.exceptionMap.containsKey(exceptionClass)) {
            return this.exceptionMap.get(exceptionClass);
        }
        Class<?> superclass = exceptionClass.getSuperclass();
        while (superclass != null) {
            if (this.exceptionMap.containsKey(superclass)) {
                ExceptionHandler matchingHandler = this.exceptionMap.get(superclass);
                this.exceptionMap.put(exceptionClass, matchingHandler); // superclass was found, avoid search next time
                return matchingHandler;
            }
            superclass = superclass.getSuperclass();
        }
        this.exceptionMap.put(exceptionClass, null); // nothing was found, avoid search next time
        return null;
    }
}

File: src/main/java/io/javalin/core/ExceptionMapper.kt
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 */

package io.javalin.core

import io.javalin.*
import io.javalin.core.util.HttpResponseExceptionMapper
import java.util.*

class ExceptionMapper {

    val exceptionMap = HashMap<Class<out Exception>, ExceptionHandler<Exception>?>()

    internal fun handle(exception: Exception, ctx: Context) {
        ctx.inExceptionHandler = true // prevent user from setting Future as result in exception handlers
        if (HttpResponseExceptionMapper.canHandle(exception) && noUserHandler(exception)) {
            HttpResponseExceptionMapper.handle(exception, ctx)
        } else {
            val exceptionHandler = this.getHandler(exception.javaClass)
            if (exceptionHandler != null) {
                exceptionHandler.handle(exception, ctx)
            } else {
                Javalin.log.warn("Uncaught exception", exception)
                HttpResponseExceptionMapper.handle(InternalServerErrorResponse(), ctx)
            }
        }
        ctx.inExceptionHandler = false
    }

    internal inline fun catchException(ctx: Context, func: () -> Unit) = try {
        func.invoke()
    } catch (e: Exception) {
        handle(e, ctx)
    }

    private fun getHandler(exceptionClass: Class<out Exception>): ExceptionHandler<Exception>? {
        if (this.exceptionMap.containsKey(exceptionClass)) {
            return this.exceptionMap[exceptionClass]
        }
        var superclass = exceptionClass.superclass
        while (superclass != null) {
            if (this.exceptionMap.containsKey(superclass)) {
                return exceptionMap[superclass]
            }
            superclass = superclass.superclass
        }
        return null
    }

    private fun noUserHandler(e: Exception) =
            this.exceptionMap[e::class.java] == null && this.exceptionMap[HttpResponseException::class.java] == null
}
